runtime/middlewares.go

```diff
@@ -0,0 +1,196 @@
+	}
+
+}
+
+// With returns a new MiddlewareStack instance that is a combination of the MiddlewareStack
+// receiver's handlers and the provided handlers.
+func (m *MiddlewareStack) With(handlers ...Handler) *MiddlewareStack {
+	return NewStack(
+		append(m.handlers, handlers...)...,
+	)
+}
+
+// Handle calls Handle on the the handlers in the stack.
+func (m *MiddlewareStack) Handle(
+	ctx context.Context,
+	req *ServerHTTPRequest,
+	res *ServerHTTPResponse) {
+	m.middleware.Handle(ctx, req, res)
+}
+
+// Use adds a Handler onto the middleware stack. Handlers are invoked in the order they are added to a MiddlewareStack.
+func (m *MiddlewareStack) Use(handler Handler) {
+	if handler == nil {
+		panic("handler cannot be nil")
+	}
+
+	m.handlers = append(m.handlers, handler)
+	m.middleware = build(m.handlers)
+}
+
+// UseFunc adds a MiddlewareStack-style handler function onto the middleware stack.
+func (m *MiddlewareStack) UseFunc(
+	handlerFunc func(
+		ctx context.Context,
+		req *ServerHTTPRequest,
+		res *ServerHTTPResponse,
+		next HandlerFn)) {
+	m.Use(HandlerFunc(handlerFunc))
+}
+
+// Handlers returns a list of all the handlers in the current MiddlewareStack.
+func (m *MiddlewareStack) Handlers() []Handler {
+	return m.handlers
+}
+
+func build(handlers []Handler) middleware {
+	if len(handlers) == 0 {
+		return voidMiddleware()
+	}
+
+	var next middleware
+	if len(handlers) > 1 {
+		next = build(handlers[1:])
+	} else {
+		next = voidMiddleware()
+	}
+
+	return middleware{handlers[0], &next}
+}
+
+func voidMiddleware() middleware {
+	return middleware{
+		HandlerFunc(
+			func(
+				ctx context.Context,
+				req *ServerHTTPRequest,
+				res *ServerHTTPResponse,
+				next HandlerFn) {
+			}),
+		&middleware{},
+	}
+}
+
+// Wrap converts a zanzibar.HandlerFn into a middleware.Handler so it can be used
+// in the middleware stack. The following zanzibar.HandlerFn is called after execution.
+func Wrap(handlerFn HandlerFn) Handler {
+	return HandlerFunc(
+		func(
+			ctx context.Context,
+			req *ServerHTTPRequest,
+			res *ServerHTTPResponse,
+			next HandlerFn) {
+			handlerFn(ctx, req, res)
+			next(ctx, req, res)
+		})
+}
+
+// UseHandlerFn adds a zanzibar.HandlerFn handler function onto the middleware stack.
+func (m *MiddlewareStack) UseHandlerFn(
+	handlerFunc func(
+		ctx context.Context,
+		req *ServerHTTPRequest,
+		res *ServerHTTPResponse)) {
+	m.Use(Wrap(handlerFunc))
+}
+
+// AdaptHTTPHandler wraps a http.Handler to support standard non-zanzibar middlewares.
+func AdaptHTTPHandler(handler http.Handler) HandlerFunc {
+	return func(
+		ctx context.Context,
+		req *ServerHTTPRequest,
+		res *ServerHTTPResponse,
+		next HandlerFn) {
+		// TODO(sindelar): Implement this with a recorder and context propogation.
+		handler.ServeHTTP(res.responseWriter, req.httpRequest)
+		next(ctx, req, res)
+	}
+}
+
+// DefaultStack returns a new middleware stack instance with the default middlewares.
+//
+// Logger - Request/Response Logging
```

Comment:
Although logger and tracer are useful examples of middlewares.

we might be better off for performance/debugging reasons to implement them as methods on the ServerReq / ServerRes class.